### PR TITLE
FLUID-5678: Added panel descriptions & panel auto-resize vertically

### PR DESCRIPTION
**Pull request by maozillah**
JIRA https://issues.fluidproject.org/browse/FLUID-5678

### CODE REVIEW

File: src/framework/preferences/html/PrefsEditorTemplate-layout.html [hunk 1]
```diff
@@ -1,11 +1,10 @@
 <h2><span class="fl-icon-toc"></span><span class="flc-prefsEditor-toc-label heading-text"></span></h2>
-
+<p class="flc-prefsEditor-toc-choice-label"></p>
```
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-layout.html]: Are these really descriptions and not labels? If they are really descriptions, the class should probably be named something related to that instead of label. Same throughout.

**Comment by jobara** [comment]
In Safari it appears that the panel heights are not lining up. Works in other browsers though.

screen shot 2015-06-17 at 2 17 57 pm

File: src/framework/preferences/html/PrefsEditorTemplate-layout.html (hunk 1 above)
Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-layout.html]: Got it, will change the class naming

**Comment by jobara** [comment]
@maozillah the new instructions should be verified in the unit tests. You can look at the panels tests, probably in the tests starting around this line https://github.com/fluid-project/infusion/blob/master/tests/framework-tests/preferences/js/PanelsTests.js#L1187

**Comment by maozillah** [comment]
What version of safari are you using justin? I'm testing it safari 8.0.6 and it seems to be working fine

**Comment by jobara** [comment]
@maozillah the safari thing was probably a cache issue. It looks good this morning.

File: src/framework/preferences/css/SeparatedPanelPrefsEditorFrame.css [hunk 2]
```diff
@@ -1 +1 @@
...
...x;float:left;}.fl-prefsEditor-separatedPanel .fl-prefsEditor-separatedPanel-toc label{line-height:1.7em !important}.fl-prefsEditor-separatedPanel .fl-links-buttons-options label{width:21em;display:inline-block}.fl-prefsEditor-separatedPanel .fl-links-buttons-options li{margin-bottom:.4em}.fl-prefsEditor-separatedPanel h2 label,.fl-prefsEditor-separatedPanel h2 .heading-text{float:left;padding-top:.2em;font-size:1.3em;font-weight:600}.fl-prefsEditor-separatedPanel .fl-icon-next,.fl-prefsEditor-separatedPanel .fl-icon-prev{font-family:'InfusionIcons' !important;speak:none;font-style:normal;font-weight:normal;font-variant:normal;text-transform:none;-webkit-font-smoothing:antialiased;float:left;margin:.2em .3em 0 0;font-size:1.5em}.fl-prefsEditor-separatedPanel .fl-icon-next:before{content:"\e001"}.fl-prefsEditor-separatedPanel .fl-icon-prev:before{content:"\e002"}.fl-prefsEditor-separatedPanel.fl-theme-bw{background-image:none}.fl-prefsEditor-separatedPanel.fl-theme-wb{background-image:none}.fl-prefsEditor-separatedPanel.fl-theme-yb{background-image:none}.fl-prefsEditor-separatedPanel.fl-theme-by{background-image:none}.fl-prefsEditor-separatedPanel.fl-theme-lgdg{background-image:none}
```
Comment (jobara) [inline src/framework/preferences/css/SeparatedPanelPrefsEditorFrame.css]: I don't think this file is supposed to be committed. @cindyli could you confirm this?

File: src/framework/preferences/html/PrefsEditorTemplate-contrast.html [hunk 3]
```diff
@@ -7,4 +8,4 @@
         <span class="fl-hidden-accessible"></span>
         <span class="fl-crossout"></span>
     </label>
-</div>
+</div>
```
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-contrast.html]: Could you please make sure that all the files have an empty new line at the end. This will remove the complaint from GitHub. There were some editors, at least historically, that would have issues with files that didn't end in an empty new line.

Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-contrast.html]: That is so particular (and such a fun fact), for sure

File: src/framework/preferences/css/SeparatedPanelPrefsEditorFrame.css (hunk 2 above)
Comment (cindyli) [inline src/framework/preferences/css/SeparatedPanelPrefsEditorFrame.css]: That's correct. All css files in src/framework/preferences/css/ directory should not be committed. They are dynamically generated at running grunt build script.

File: tests/framework-tests/preferences/js/PanelsTests.js [hunk 4]
```diff
@@ -1187,9 +1201,17 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
...
-    fluid.tests.testDefault = function (that) {
+    fluid.tests.testDefault = function (that, messageMap) {
         var inputValue = that.container.find("input").val();
+        var messageBase = that.options.messageBase;
         jqUnit.assertEquals("The default input value has been set to the min value", that.options.range.min, inputValue);
+
+        fluid.each(messageMap, function(messageName, selectorName) {
+            jqUnit.assertEquals("The label text is " + messageBase[messageName], messageBase[messageName], that.locate(selectorName).text());
+        //     jqUnit.assertEquals("The label text is " + messageBase.textSizeLabel, messageBase.textSizeLabel, that.locate("label").text());
+        // jqUnit.assertEquals("The description text is " + messageBase.textSizeDescr, messageBase.textSizeDescr, that.locate("textSizeDescr").text());
```
Comment (jobara) [inline tests/framework-tests/preferences/js/PanelsTests.js]: the commented out code should be removed.

File: src/framework/preferences/html/PrefsEditorTemplate-speak.html [hunk 5]
```diff
@@ -1,11 +1,10 @@
 <h2><span class="fl-icon-speak"></span><span class="flc-prefsEditor-speak-label heading-text"></span></h2>
-
+<p class="flc-prefsEditor-speak-choice-label"></p>
 <div class="fl-prefsEditor-onoff">
     <input type="checkbox" id="speak-choice" class="flc-prefsEditor-speak" />
     <label for="speak-choice">
         <span class="fl-prefsEditor-switch" data-checkboxStateOn="ON" data-checkboxStateOff="OFF">
             <span class="fl-prefsEditor-switch-inner"></span>
         </span>
-        <span class="flc-prefsEditor-speak-choice-label"></span>
```
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: The problem with this change is that there is no longer a label for the checkbox. When an assistive technology comes to the checkbox it will only read its state (checked/unchecked) but not what it is for. We need to maintain a label.

Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: I think this is out of scope for this fix and will be address with https://issues.fluidproject.org/browse/FLUID-5417

File: src/framework/preferences/messages/linksControls.json [hunk 6]
```diff
@@ -1,3 +1,4 @@
 {
-    "linksControlsLabel": "Links & buttons"
+    "emphasisDescr": "Emphasize input options",
```
Comment (jobara) [inline src/framework/preferences/messages/linksControls.json]: Were all of the text changes vetted by @dayotte?

Comment (maozillah) [inline src/framework/preferences/messages/linksControls.json]: Unofficially, they were in the design mock for suggested UIO updates https://www.dropbox.com/s/02j7ilnhxd7c8hv/UIO_mock.pdf?dl=0

Comment (dayotte) [inline src/framework/preferences/messages/linksControls.json]: I think I need to walk through this with you @maozillah and/or @jobara to understand what all the changes are.

File: src/framework/preferences/html/PrefsEditorTemplate-speak.html (hunk 5 above)
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: In the original code, the span is included in the label, so it could be read by an AT. However, in the change, there is nothing provided for a label at all.

Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: Gotcha, will make the change

File: src/framework/preferences/messages/linksControls.json (hunk 6 above)
Comment (dayotte) [inline src/framework/preferences/messages/linksControls.json]: @jobara: talked to @maozillah about this and she will make some changes based on our discussion

File: src/framework/preferences/html/PrefsEditorTemplate-speak.html [hunk 7]
```diff
@@ -1,11 +1,10 @@
 <h2><span class="fl-icon-speak"></span><span class="flc-prefsEditor-speak-label heading-text"></span></h2>
-
 <div class="fl-prefsEditor-onoff">
     <input type="checkbox" id="speak-choice" class="flc-prefsEditor-speak" />
     <label for="speak-choice">
+    	<p class="flc-prefsEditor-speak-choice-label"></p>
```
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: `<p>` tags are not valid descendants of a `<label>`. Basically you can't use block level elements inside of an inline element. You can use this validator to check the markup, but you don't need to worry about things that will be adjusted by the javascript (e.g. the empty heading).

https://validator.w3.org/nu/

File: src/framework/preferences/messages/speak.json [hunk 8]
```diff
@@ -1,4 +1,4 @@
 {
     "speakLabel": "Text-to-Speech",
-    "speakChoiceLabel": "Enable the page to speak itself out loud."
+    "speakDescr": "Let computer read site content out loud"
```
Comment (jobara) [inline src/framework/preferences/messages/speak.json]: Maybe this should be "Let the computer read site content out loud"

File: src/framework/preferences/js/Panels.js [hunk 9]
```diff
@@ -916,10 +924,10 @@ var fluid_2_0 = fluid_2_0 || {};
         },
         selectors: {
             links: ".flc-prefsEditor-links",
-            linksChoiceLabel: ".flc-prefsEditor-links-choice-label"
+            linkschoiceLabel: ".flc-prefsEditor-links-choice-label"
```
Comment (jobara) [inline src/framework/preferences/js/Panels.js]: Was there a reason to change the name from "linksChoiceLabel" to "linkschoiceLabel"? Similar for "inputschoiceLabel".

File: src/framework/preferences/html/PrefsEditorTemplate-speak.html (hunk 7 above)
Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: Wanted to have the label and toggle button on two different lines, changed it to a span class but added two <br> after. not sure if that's best practice given that it's an inline label element...? Perhaps this will be addressed in the other jira for moving to fieldset

File: src/framework/preferences/html/PrefsEditorTemplate-layout.html [hunk 10]
```diff
@@ -1,11 +1,10 @@
 <h2><span class="fl-icon-toc"></span><span class="flc-prefsEditor-toc-label heading-text"></span></h2>
-
 <div class="fl-prefsEditor-onoff">
     <input type="checkbox" id="toc-choice" class="flc-prefsEditor-toc" />
     <label for="toc-choice">
+    	<span class="flc-prefsEditor-toc-descr"></span><br><br>
```
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-layout.html]: would you be to accomplish the effects of using <br> in css instead?

File: src/framework/preferences/html/PrefsEditorTemplate-speak.html (hunk 7 above)
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: I left a comment about this below. Also if you use html tags straight in the comments here, they will be treated as actual markup, because these comment fields support markdown. You should wrap the tag name in ` to have it styled as a code. You can look at the [GitHub Flavoured Markdown](https://help.github.com/articles/github-flavored-markdown/) page for more info on syntax.

Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-speak.html]: Ahh gotcha, thanks for the headsup! and yep I'll work on addressing that in CSS

File: demos/prefsFramework/html/SimplifyPanelTemplate.html [hunk 11]
```diff
@@ -1,11 +1,10 @@
 <h2><span class="fl-icon-simplify"></span><span class="demo-prefsEditor-simplify-label heading-text"></span></h2>
-
 <div class="fl-prefsEditor-onoff">
     <input type="checkbox" id="simplify-choice" class="demo-prefsEditor-simplify" />
     <label for="simplify-choice">
+    	<span class="demo-prefsEditor-simplify-choice-label block"></span>
```
Comment (jobara) [inline demos/prefsFramework/html/SimplifyPanelTemplate.html]: the styling classes should be in a fl- namespace if they are part of the component or a demo- namespace if they are part of the demo.

File: src/framework/preferences/css/SeparatedPanelPrefsEditorFrame.css [hunk 12]
```diff
@@ -1 +1 @@
...
```
Comment (jobara) [inline src/framework/preferences/css/SeparatedPanelPrefsEditorFrame.css]: this file and any other css file generated from stylus should be removed from the repo. Only the stylus files should be committed.

File: src/framework/preferences/css/stylus/SeparatedPanelPrefsEditorFrame.styl [hunk 13]
```diff
@@ -53,13 +54,20 @@ html {
...
-            width: 25em;
+            min-width: 25em;
             background-color: #fff;
-            margin-right: 10px;
             white-space: normal;
-            vertical-align: middle;
+            display: table-cell;
         }
+        
+        .block {
```
Comment (jobara) [inline src/framework/preferences/css/stylus/SeparatedPanelPrefsEditorFrame.styl]: this should be in an fl-prefsEditor- namespace. It would also be good to know what it is used for. Maybe a comment or something. Also, what are the -webkit- prefixed styles for, why not just `display: block`?

File: src/framework/preferences/css/stylus/SeparatedPanelPrefsEditorFrame.styl [hunk 14]
```diff
@@ -53,13 +54,17 @@ html {
...
+            min-width: 25em;
             background-color: #fff;
-            margin-right: 10px;
             white-space: normal;
-            vertical-align: middle;
+            display: table-cell;
         }
+        
+        // override inline label style for panel description 
+        .fl-prefsEditor-display-block {
```
Comment (jobara) [inline src/framework/preferences/css/stylus/SeparatedPanelPrefsEditorFrame.styl]: instead of making this special class, i wonder if we just add styling for an `fl-prefsEditor-toc-descr` class. e.g.

```css
label .fl-prefsEditor-toc-descr {
    display: block; 
    margin-bottom: 1em;
}
```

File: src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html [hunk 15]
```diff
@@ -1,5 +1,6 @@
 <div class="flc-prefsEditor-line-space">
     <h2><span class="fl-icon-line-space"></span><label class="flc-prefsEditor-line-space-label" for="line-space"></label></h2>
```
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html]: If the `<label>` in the header of the panel are not actually being used as labels, could you change them to be `<span>`.

File: src/framework/preferences/css/stylus/SeparatedPanelPrefsEditorFrame.styl (hunk 14 above)
Comment (maozillah) [inline src/framework/preferences/css/stylus/SeparatedPanelPrefsEditorFrame.styl]: Great idea, I added the styling to the on/off toggle instead which will make it easier to scale in the future as well

File: src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html (hunk 15 above)
Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html]: This greatly messes the styling. Wondering if it would be okay to leave them in for now as they will probably be used when more semantic descriptions are added for this jira https://issues.fluidproject.org/browse/FLUID-5417

Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html]: Currently the description of [FLUID-5678](https://issues.fluidproject.org/browse/FLUID-5678) discusses addressing three things.

- Detailed description/instruction text for function of tools e.g. change size of site text  
- Use of labels on the headings is inconsistent – some headings are labels for a panel’s control and some are not. Input identification/labeling should be consistent   
- Wording for measure of scale with text size and line height as x Times may be confusing 

If you'd like to do the labels and etc. in other pull requests, that's okay, but please create new jiras for them and update the description for this one.

**Comment by jobara** [comment]
@maozillah it seems there are merge conflicts, likely from changes in https://github.com/fluid-project/infusion/pull/615 that was just merged. Please update this pull request from master to address the conflicts.

File: src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html (hunk 15 above)
Comment (maozillah) [inline src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html]: Updated [https://issues.fluidproject.org/browse/FLUID-5417]FLUID-5417 description to include the label task and [https://issues.fluidproject.org/browse/FLUID-5678]FLUID-5678

**Comment by maozillah** [comment]
@jobara Reviewed and updated merge conflicts. This branch should overwrite the two existing conflicts leftover for "colour and contrast" label and number of unit tests for the contrast panel

File: src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html (hunk 15 above)
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-lineSpace.html]: Thanks.

**Comment by jobara** [comment]
@maozillah looks like there's still a merge conflict

**Comment by jobara** [comment]
For the full page examples, the description seems to be a bit misaligned. Would you be able to address that as well? (the full page prefs editors can be found in the manual tests).

screen shot 2015-07-14 at 2 28 55 pm

**Comment by jobara** [comment]
@maozillah there seems to be a bug with prefs framework examples in the examples directory. As you toggle the on/off adjusters, the panels creep lower and lower down the page.

screen shot 2015-07-14 at 2 34 04 pm

**Comment by maozillah** [comment]
Moved the description before the toggle to address the issue for now and left a comment. It will be addressed in [FLUID-5708](https://issues.fluidproject.org/browse/FLUID-5708)

**Comment by jobara** [comment]
@maozillah could you also provide an accompanying pull request again [infusion-docs](https://github.com/fluid-project/infusion-docs). At least the following should be updated, but you should probably double check the rest of the Prefs Framework and UIO docs as well:

* [Creating a New Adjuster UI](http://docs.fluidproject.org/infusion/development/tutorial-creatingANewAdjusterUI/CreatingANewAdjusterUI.html)
* [API Changes From 1.5 to 2.0](http://docs.fluidproject.org/infusion/development/APIChangesFrom1_5To2_0.html)

**Comment by maozillah** [comment]
Issued a new [pull request](https://github.com/fluid-project/infusion-docs/pull/71) on the infusion docs including updates from this pull request.

Checked out API changes but wasn't sure what to add.

**Comment by maozillah** [comment]
Updated some selectors for consistency

File: src/framework/preferences/js/Panels.js [hunk 16]
```diff
@@ -660,7 +660,8 @@ var fluid_2_0 = fluid_2_0 || {};
             label: ".flc-prefsEditor-min-text-size-label",
             smallIcon: ".flc-prefsEditor-min-text-size-smallIcon",
             largeIcon: ".flc-prefsEditor-min-text-size-largeIcon",
-            multiplier: ".flc-prefsEditor-multiplier"
+            multiplier: ".flc-prefsEditor-multiplier",
+            textSizeDescr: ".flc-prefsEditor-textsize-descr"
```
Comment (jobara) [inline src/framework/preferences/js/Panels.js]: There is a naming inconsistency with the other class names. `".flc-prefsEditor-textsize-descr"` vs `".flc-prefsEditor-min-text-size-largeIcon"` We should keep the class names consistent.

Comment (jobara) [inline src/framework/preferences/js/Panels.js]: There are similar issues throughout. Keep in mind the camelCasing and hyphenation used.

File: src/framework/preferences/html/PrefsEditorTemplate-linksControls.html [hunk 17]
```diff
@@ -1,4 +1,5 @@
 <h2><span class="fl-icon-links"></span><span class="flc-prefsEditor-linksControls-label heading-text"></span></h2>
+<p class="flc-prefsEditor-emphasis-descr"></p>
```
Comment (jobara) [inline src/framework/preferences/html/PrefsEditorTemplate-linksControls.html]: This should probably have been `flc-prefsEditor-emphasizeLinks-descr`, but i'm not sure where it is defined/used.

**Comment by maozillah** [comment]
Updated

**Comment by jobara** [comment]
@maozillah in Safari, the focus styling for the table of contents adjuster only goes around the label. It should go around both the control and lable. This is working in Firefox and Chrome.

screen shot 2015-08-06 at 9 33 19 am

**Comment by jobara** [comment]
seems to work nicely on the full page version though.

screen shot 2015-08-06 at 10 04 18 am

**Comment by maozillah** [comment]
Fixed

**Comment by jobara** [comment]
@maozillah just noticed that you've included the .DS_Store file. That's a Mac OS file that should be included in your .gitignore. Please remove it from the pull request too.

**Comment by maozillah** [comment]
Deleted

**Comment by jobara** [comment]
@maozillah is it possible to have all of the descriptions start with a capital?

screen shot 2015-08-12 at 10 10 20 am

**Comment by maozillah** [comment]
Fixed

**Comment by jobara** [comment]
@amb26 this pull request is ready to be merged. Because this will go in after https://github.com/fluid-project/infusion/pull/591, would you be able to help fix any merge conflicts.

**Comment by jobara** [comment]
Closed in favour of https://github.com/fluid-project/infusion/pull/639 which uses the contents of this pull request but is updated to the latest master with the conflicts fixed.
